package-res/resources/web/pentaho/type/_type.js

```diff
@@ -0,0 +1,1362 @@
+       *
+       * @return {!pentaho.type.spec.ITypeProto} A specification of this type.
+       *
+       * @see pentaho.type.Type#toSpecInContext
+       * @see pentaho.type.Type#_fillSpecInContext
+       */
+      toSpec: function(keyArgs) {
+        return O.using(new SpecificationScope(), this.toSpecInContext.bind(this, keyArgs || {}));
+      },
+
+      /**
+       * Creates a specification that describes this type.
+       *
+       * This method requires that there currently exists an
+       * [ambient specification context]{@link pentaho.type.SpecificationContext.current}.
+       *
+       * The default implementation returns a plain object with the id of the type and
+       * any other specified attributes
+       * (like [label]{@link pentaho.type.Type#label} or [description]{@link pentaho.type.Type#description}).
+       *
+       * @param {Object} [keyArgs] - The keyword arguments object.
+       * Passed to every type and instance serialized within this scope.
+       *
+       * Please see the documentation of subclasses for information on additional, supported keyword arguments.
+       *
+       * @return {!pentaho.type.spec.ITypeProto} A specification of this type.
+       *
+       * @see pentaho.type.Type#toSpec
+       */
+      toSpecInContext: function(keyArgs) {
+
+        var spec = {id: this.shortId || SpecificationContext.current.add(this)};
+
+        this._fillSpecInContext(spec, keyArgs || {});
+
+        return spec;
+      },
+
+      /**
+       * Creates a JSON specification that describes this type.
+       *
+       * Attributes which don't have a JSON-compatible specification are omitted.
+       * Specifically, attributes with a function value are not supported.
+       *
+       * This method simply calls {@link @see pentaho.type.Instance#toSpec} with argument `keyArgs.isJson` as `true`
+       * and exists for seamless integrations with JavaScript's
+       * [JSON.stringify](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)
+       * method.
+       *
+       * @see pentaho.type.Instance#toSpec
+       *
+       * @return {UJsonValue} A JSON-compatible specification.
+       */
+      toJSON: function() {
+        return this.toSpec({isJson: true});
+      },
+
+      /**
+       * Fills the given specification with this type's attributes' local values,
+       * and returns whether any attribute was actually added.
+       *
+       * This method requires that there currently exists an
+       * [ambient specification context]{@link pentaho.type.SpecificationContext.current}.
+       *
+       * This method does _not_ add the special `id` and `base` attributes to the specification.
+       *
+       * @param {!Object} spec - The specification to be filled.
+       * @param {Object} [keyArgs] - The keyword arguments object.
+       * Passed to every type and instance serialized within this scope.
```

Comment:
Suggest rewording into a sentence structure:

The keyword arguments object is passed to every type and instance serialized within this scope.